vignettes/opencl.Rmd

```diff
@@ -0,0 +1,143 @@
+---
+title: "Running Stan on the GPU with OpenCL"
+author: "Rok Češnovar and Jonah Gabry"
+output: 
+  rmarkdown::html_vignette:
+    toc: true
+    toc_depth: 4
+params:
+  EVAL: !r identical(Sys.getenv("CMDSTANR_OPENCL_TESTS"), "true")
+vignette: >
+  %\VignetteIndexEntry{Running Stan on the GPU with OpenCL}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## Introduction
+
+This vignette demonstrates how to use the OpenCL capabilities of CmdStan with
+CmdStanR. The functionality described in this vignette requires CmdStan 2.26.1
+or newer.
+
+As of version 2.26.1, users can expect speedups with OpenCL when using vectorized
+probability distribution functions (functions with the `_lpdf` or `_lpmf`
+suffix) and when the input variables contain at least 20,000 elements.
+
+The actual speedup for a model will depend on the particular `lpdf/lpmf`
+functions used and whether the `lpdf/lpmf` functions are the bottlenecks of the
+model. The more computationally complex the function is, the larger the expected
+speedup. The biggest speedups are expected when using the specialized GLM
+functions.
+
+In order to establish the bottlenecks in your model we recommend using
+[profiling](../profiling.Rmd), which was introduced in Stan version 2.26.0. 
+
+## OpenCL runtime
+
+OpenCL is supported on most modern CPUs and GPUs. In order to use
+OpenCL in CmdStanR, an OpenCL runtime for the target device must be installed.
+A guide for the most common devices is available in the CmdStan manual's 
+[chapter on parallelization](https://mc-stan.org/docs/2_26/cmdstan-guide/parallelization.html#opencl).
+
+## Compiling a model with OpenCL
+
+By default, models in CmdStanR are compiled *without* OpenCL support. Once OpenCL
+support is enabled, a CmdStan model will make use of OpenCL if the functions
+in the model support it. Technically no changes to a model are required to
+support OpenCL since the choice of using OpenCL is handled by the compiler,
+but it can still be useful to rewrite a model to be more OpenCL-friendly by
+using vectorization as much as possible when using probability distributions.
+
+Consider a simple logistic regression with parameters `alpha` and `beta`,
+covariates `X`, and outcome `y`.
+
+```
+data {
+  int<lower=1> k;
+  int<lower=0> n;
+  matrix[n, k] X;
+  int y[n];
+}
+parameters {
+  vector[k] beta;
+  real alpha;
+}
+model {
+  target += std_normal_lpdf(beta);
+  target += std_normal_lpdf(alpha);
+  target += bernoulli_logit_glm_lpmf(y | X, alpha, beta);
+}
+```
+
+Some fake data will be useful to run this model:
+
+```{r fake-data, message=FALSE, results='hide'}
+library(cmdstanr)
+
+# Generate some fake data
+n <- 200000
+k <- 20
+X <- matrix(rnorm(n * k), ncol = k)
+y <- rbinom(n, size = 1, prob = plogis(3 * X[,1] - 2 * X[,2] + 1))
```

Comment:
I switched to using `rbinom()` here, which is a bit more natural in R. Also `plogis()`avoids having to write out the math of the inverse logit function.

Reply:
Cool, makes sense. Will rerun the vignette locally.